frontend/ui/lib/widgets/retryable_async_loadable.dart

```diff
@@ -0,0 +1,88 @@
+import 'package:flutter/material.dart';
+
+/// Provides loading and retrying functionality.
+/// This should only be used with an idempotent async function.
+class RetryableAsyncLoadable<T> extends StatefulWidget {
+  final _OnLoad<T> onLoad;
+  final _LoadFunction<T> loadFunction;
+  final String errorMessage;
+
+  RetryableAsyncLoadable({
+    key,
+    @required this.onLoad,
+    @required this.loadFunction,
+    @required this.errorMessage,
+  }) : super(key: key);
+
+  @override
+  State createState() =>
+      _RetryableAsyncLoadableState(onLoad, loadFunction, errorMessage);
+}
+
+typedef _LoadFunction<T> = Future<T> Function();
+typedef _OnLoad<T> = Widget Function(T);
+
+class _RetryableAsyncLoadableState<T> extends State<RetryableAsyncLoadable<T>> {
+  final _OnLoad<T> onLoad;
+  final _LoadFunction<T> loadFunction;
+  final String errorMessage;
+  bool loading = true;
+  Exception error;
+  T t;
+
+  _RetryableAsyncLoadableState(
+      this.onLoad, this.loadFunction, this.errorMessage) {
+    preformAction();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    if (loading) {
+      return Center(child: CircularProgressIndicator());
+    }
+    if (error != null) {
+      return AlertDialog(
+        title: Text(errorMessage),
+        content: Text('Would you like to retry?'),
+        actions: [
+          FlatButton.icon(
+            onPressed: () {
+              preformAction();
+            },
+            icon: Icon(Icons.refresh),
+            label: Text('Retry'),
+          ),
+        ],
+      );
+    }
+    if (t != null) {
+      return onLoad.call(t);
+    }
+    // TODO: do something smart here...
+    return Container();
+  }
+
+  void preformAction() async {
```

Comment:
oh, yeah this should be perform. And no need for the underscore because the class itself is already file-private. This is not something that is overridden, it's just like, refresh the widget and restart the load.

Reply:
renamed it to `startLoad`